operators/distiller-pipeline/distiller_pipeline/__init__.py

```diff
@@ -0,0 +1,981 @@
+        return cls()
+
+
+class NodeInfo(BaseModel):
+    n_messages: int = 0
+    node_id: str = "0"
+    node_group_thread_info: dict[int, dict[int, dict[int, NodeGroupThreadInfo]]] = (
+        Field(default_factory=dict)
+    )
+
+    @classmethod
+    def from_msgpack(cls, data: list[Any]) -> "NodeInfo":
+        instance_data: dict[str, Any] = {
+            "n_messages": 0,
+            "node_id": "0",
+            "node_group_thread_info": {},
+        }
+        if isinstance(data, list) and len(data) >= 3:
+            try:
+                instance_data["n_messages"] = int(data[0])
+            except (ValueError, TypeError):
+                pass
+
+            node_id_raw = data[1]
+            try:
+                instance_data["node_id"] = (
+                    node_id_raw.decode("utf-8")
+                    if isinstance(node_id_raw, bytes)
+                    else str(node_id_raw)
+                )
+            except UnicodeDecodeError:
+                instance_data["node_id"] = str(node_id_raw)
+
+            if isinstance(data[2], dict):
+                try:
+                    instance_data["node_group_thread_info"] = {
+                        int(mpi_rank): {
+                            int(short_id): {
+                                int(thread_id): NodeGroupThreadInfo.from_msgpack(
+                                    info_data,
+                                )
+                                for thread_id, info_data in threads_map.items()
+                            }
+                            for short_id, threads_map in short_ids_map.items()
+                        }
+                        for mpi_rank, short_ids_map in data[2].items()
+                    }
+                except (ValueError, TypeError):
+                    instance_data["node_group_thread_info"] = {}
+
+        try:
+            return cls(**instance_data)
+        except ValidationError:
+            return cls()
+
+
+class SharedState(BaseModel):
+    receiver_map: dict[str, ReceiversInfo] = Field(default_factory=dict)
+    aggregator_map: dict[str, AggregatorsInfo] = Field(default_factory=dict)
+    node_map: dict[str, NodeInfo] = Field(default_factory=dict)
+
+    @classmethod
+    def from_msgpack(cls, data: list[Any]) -> "SharedState":
+        instance_data: dict[str, Any] = {
+            "receiver_map": {},
+            "aggregator_map": {},
+            "node_map": {},
+        }
+        if isinstance(data, list) and len(data) >= 3:
+            if isinstance(data[0], dict):
+                try:
+                    instance_data["receiver_map"] = {
+                        (
+                            k.decode("utf-8") if isinstance(k, bytes) else str(k)
+                        ): ReceiversInfo.from_msgpack(v)
+                        for k, v in data[0].items()
+                    }
+                except (UnicodeDecodeError, ValueError, TypeError):
+                    instance_data["receiver_map"] = {}
```

Comment:
This whole block looks like it duplicated a couple of times? Maybe pull out into method?

Reply:
agree, I did try this refactor but there are slight differences in the msgpack types. I will improve it.